### PR TITLE
Run virt-launcher as spc_t for failing multiqueue tests

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
This should fix test 1665 on the k8s-1.17 lane.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by jean-edouard** [comment]
Test added

File: pkg/virt-controller/services/template_test.go [hunk 1]
```diff
@@ -203,6 +203,27 @@ var _ = Describe("Template", func() {
...
+								NetworkInterfaceMultiQueue: &enabled,
+							},
+						},
+					},
+				}
+				pod, err := svc.RenderLaunchManifest(&vmi)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(pod.Spec.SecurityContext).ToNot(BeNil())
+				Expect(pod.Spec.SecurityContext.SELinuxOptions).ToNot(BeNil())
+				Expect(pod.Spec.SecurityContext.SELinuxOptions.Type).To(Equal("spc_t"))
```
Comment (vladikr) [inline pkg/virt-controller/services/template_test.go:225]: One thing that I'm concerned about is that the previous test on line 190, sets the spc_t in the fake cluster config, but never removes it. Perhaps you can clear it at the end of the previous test so we'll know that this one is 100% correct.

Comment (jean-edouard) [reply pkg/virt-controller/services/template_test.go:228]: Yes, fixed! This is a problem in other places as well, which I used as a model...

**Comment by vladikr** [comment]
Both of these options will set spc_t for all vmis in the cluster... we will need to find another way.

**Comment by rmohr** [comment]
I would be ok with going with spc_t in the context of this PR, since it would at least conserve the progress which we have on all other areas. If we can't make the selinux k8s lane green, then we can't protect us against regressions.

I of course agree with @fabiand that this is not a solution where we would say that we solved a the security problem.

**Comment by davidvossel** [comment]
> I would be ok with going with spc_t in the context of this PR, since it would at least conserve the progress which we have on all other areas. If we can't make the selinux k8s lane green, then we can't protect us against regressions.

before we consider merging something like this, which has pretty serious security implications even if it is limted and temporary, do we have any other options?  Basically i want to make sure we've exhausted all options before considering something like this...

My understanding from talking with you all yesterday out of band is that spc_t is required because of some sort of logic libvirt is attempting to perform.  Is it possible to perform this logic in virt-handler, and then use qemu args directly to bypass libvirt?

**Comment by rmohr** [comment]
> before we consider merging something like this, which has pretty serious security implications even if it is limted and temporary, do we have any other options? Basically i want to make sure we've exhausted all options before considering something like this...

Oh wait, we don't use spc_t right now. Let's not do that. I was on the completely wrong track. I thought that we would not make it worse than before, but we actually would. Right?

**Comment by rmohr** [comment]
sorry, wrong PR.

**Comment by jean-edouard** [comment]
@rmohr as you said, the only point of this PR would be a temporary fix to make 1.17 green, so 1.17 can be enabled in github and we can ensure new PRs don't introduce new denials.

**Comment by rmohr** [comment]
> @rmohr as you said, the only point of this PR would be a temporary fix to make 1.17 green, so 1.17 can be enabled in github and we can ensure new PRs don't introduce new denials.

ok, so let's do the following: put a beforeEach and afterEach to it, which configure kubevirt-config to use `spc_t`. We have a test in `tests/security_features_test.go` which shows how to do that.

This way we can run the tests and continue testing the codepaths without introducing any risks.

**Comment by jean-edouard** [comment]
Good idea! Will open a new PR then. Closing this. Thanks!

**Comment by vladikr** [comment]
@jean-edouard Personally, I don't think you need to open and close PRs just because the direction has changed. I think it's even better to continue working in a single PR so that it's easier to keep track of the decisions. But that's just my take.

**Comment by vladikr** [comment]
> My understanding from talking with you all yesterday out of band is that spc_t is required because of some sort of logic libvirt is attempting to perform. Is it possible to perform this logic in virt-handler, and then use qemu args directly to bypass libvirt?

@davidvossel I think it's very hard to do with the current approach. Libvirt just has to create a tap device with multiple sockets in the context of qemu. I hope we will be able to add the following permissions to container_t policy: ```tun_socket { attach_queue relabelfrom relabelto create_socket_perms } ```

**Comment by maiqueb** [comment]
> > My understanding from talking with you all yesterday out of band is that spc_t is required because of some sort of logic libvirt is attempting to perform. Is it possible to perform this logic in virt-handler, and then use qemu args directly to bypass libvirt?
> 
> @davidvossel I think it's very hard to do with the current approach. Libvirt just has to create a tap device with multiple sockets in the context of qemu. I hope we will be able to add the following permissions to container_t policy: `tun_socket { attach_queue relabelfrom relabelto create_socket_perms } `

@vladikr Could you elaborate a bit here ? I'm interested in that for #3290 . Is that request / work being tracked somewhere ?

**Comment by jean-edouard** [comment]
Before I force-push over this whole PR, does this look like what we want? https://github.com/jean-edouard/kubevirt/commit/633b107a66d75f911fce909428c199b903380c79

File: tests/vmi_multiqueue_test.go [hunk 2]
```diff
@@ -20,22 +20,63 @@
...
+	Expect(err).ToNot(HaveOccurred())
+	originalContext, present := kubeVirtConfig.Data[virtconfig.SELinuxLauncherTypeKey]
+	kubeVirtConfig.Data[virtconfig.SELinuxLauncherTypeKey] = "spc_t"
+	newData, err := json.Marshal(kubeVirtConfig.Data)
+	Expect(err).ToNot(HaveOccurred())
+	data := fmt.Sprintf(`[{ "op": "replace", "path": "/data", "value": %s }]`, string(newData))
+	kubeVirtConfig, err = (*virtClient).CoreV1().ConfigMaps(tests.KubeVirtInstallNamespace).Patch("kubevirt-config", types.JSONPatchType, []byte(data))
+	Expect(err).ToNot(HaveOccurred())
+	// Allow time for virt-controller's ConfigMap cache to sync
+	time.Sleep(3 * time.Second)
```
Comment (vladikr) [inline tests/vmi_multiqueue_test.go:54]: I think yon replace this with 
```suggestion
tests.UpdateClusterConfigValueAndWait(virtconfig.SELinuxLauncherTypeKey, string(newData)
```

File: tests/vmi_multiqueue_test.go [hunk 3]
```diff
@@ -20,22 +20,63 @@
...
 
 	v1 "kubevirt.io/client-go/api/v1"
 	"kubevirt.io/client-go/kubecli"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
 	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
 	"kubevirt.io/kubevirt/tests"
 )
 
+func setSpcT(virtClient *kubecli.KubevirtClient) *string {
+	kubeVirtConfig, err := (*virtClient).CoreV1().ConfigMaps(tests.KubeVirtInstallNamespace).Get("kubevirt-config", metav1.GetOptions{})
```
Comment (vladikr) [inline tests/vmi_multiqueue_test.go:44]:
```suggestion
	kubeVirtConfig, err := tests.GetKubeVirtConfigMap()
```

File: tests/vmi_multiqueue_test.go (hunk 2 above)
Comment (jean-edouard) [reply tests/vmi_multiqueue_test.go:54]: Thank you, I knew there had to be a simpler way!
Will fix.

**Comment by jean-edouard** [comment]
Force-pushed a much simpler approach

**Comment by jean-edouard** [comment]
This PR can be closed if/when https://github.com/kubevirt/kubevirt/pull/3371 is approved.

**Comment by jean-edouard** [comment]
Superseded by https://github.com/kubevirt/kubevirt/pull/3371. Closing

**Pull request closed without merging**